File: python/space-age/space_age.py
__author__ = 'Cedric Zhuang'


class SpaceAge(object):
    RATIO_MAP = {
        "earth": 1.0,
        "mercury": 0.2408467,
        "venus": 0.61519726,
        "mars": 1.8808158,
        "jupiter": 11.862615,
        "saturn": 29.447498,
        "uranus": 84.016846,
        "neptune": 164.79132
    }

    EARTH = 31557600

    def __init__(self, seconds):
        self.seconds = seconds

    def __getattr__(self, item):
        planet = ''
        if item.startswith('on_'):
            planet = item[3:]

        if planet in self.RATIO_MAP.keys():
            ratio = self.RATIO_MAP[planet]
            ret = lambda: round(self.seconds / ratio / self.EARTH, 2)
        else:
            raise NotImplementedError('attribute is not defined.')

        return ret

File: cpp/space-age/space_age.cpp
#include "space_age.h"
#include <map>

namespace space_age
{

const long long EARTH_YEAR = 31557600;

const std::map<PLANETS, double> YEAR_MAPPING{
    {earth, 1},
    {mercury, 0.2408467},
    {venus, 0.61519726},
    {mars, 1.8808158 },
    {jupiter, 11.862615},
    {saturn, 29.447498 },
    {uranus, 84.016846 },
    {neptune, 164.79132 }
};

space_age::space_age(const long long& seconds) : _seconds(seconds) {}

const long long space_age::seconds() const
{
    return _seconds;
}

const double space_age::_to_planet(PLANETS planet) const
{
    return (double)_seconds / (double)EARTH_YEAR / YEAR_MAPPING.at(planet);
}

} // space_age